url416ee5b4ee_martinmarinov_HonProj_tgz-pJ8-martin_gui_MainJ8/src/martin/gui/quantum/Qubit.java

```diff
@@ -0,0 +1,421 @@
+	public void mouseClick(Graphics2D g, int x, int y, Visualizer vis) {
+		final int[] temp = new int[] {x, y};
+		snapToGrid(temp);
+		
+		vis.addItem(new Qubit(t, temp[0], temp[1], getFirstFreeId(vis)));
+		vis.grabDefaultTool();
+	}
+	
+	private boolean isThereAQubitWithId(int id, Visualizer vis) {
+		for (final Item i : vis.items)
+			if (i instanceof Qubit && ((Qubit) i).id == id)
+				return true;
+		
+		return false;
+	}
+	
+	private int getFirstFreeId(Visualizer vis) {
+		int id = 1;
+		while (isThereAQubitWithId(id, vis)) id++;
+		return id;
+	}
+
+	@Override
+	public void mouseDrag(Graphics2D g, int x, int y, Visualizer vis) {
+		mouseMove(g, x, y, vis);
+	}
+
+	@Override
+	String getIconFileName() {
+		return ICON + "_" + t + ".png";
+	}
+
+	@Override
+	void mousePressed(Graphics2D g, int x, int y, Visualizer vis) {
+		mouseClick(g, x, y, vis);
+	}
+
+	@Override
+	void mouseReleased(Graphics2D g, int x, int y, Visualizer vis) {
+		mouseMove(g, x, y, vis);
+	}
+
+	@Override
+	String[] getMenuEntries(Visualizer vis) {
+		final type[] types = type.values();
+		final ArrayList<String> entries = new ArrayList<String>();
+		menu_types = new type[types.length - 1];
+		
+		int j = 0;
+		for (int i = 0; i < types.length; i++)
+			if (types[i] != t) {
+				entries.add("Change to "+types[i]+" qubit");
+				menu_types[j++] = types[i];
+			}
+		
+		// additional actions, starting from id 0
+		int additional_id = 0;
+		menu_swap_id = -1;
+		menu_measurement_angle = -1;
+		menu_perform_measurement = -1;
+		
+		if (numberOfQubitsInVisualizer(vis) > 1) {
+			entries.add("Edit qubit id");
+			menu_swap_id = additional_id++;
+		}
+		
+		entries.add((perform_measurement ? "â " : "") + "Perform measurement");
+		menu_perform_measurement = additional_id++;
+
+		if (perform_measurement) {
+			entries.add("Set measurement angle");
+			menu_measurement_angle = additional_id++;
+		}
+
+		return entries.toArray(new String[0]);
+	}
+	
+	private int numberOfQubitsInVisualizer(final Visualizer vis) {
+		int c = 0;
+		
+		for (final Item i : vis.items)
+			if (i instanceof Qubit)
+				c++;
+		
+		return c;
+	}
+	
+	public type getType() {
+		return t;
+	}
+	
+	/**
+	 * Get 0 based qubit id. Keep in mind this is 0 based!
+	 * @return
+	 */
+	public int getId() {
+		return id - 1;
+	}
+	
+	public Qubit getHighestQubitId(final Visualizer vis) {
+		int highest = -1;
+		Qubit ans = null;
+		
+		for (final Item i : vis.items)
+			if (i instanceof Qubit) {
+				final Qubit q = (Qubit) i;
+				if (q.id > highest) {
+					highest = q.id;
+					ans = q;
+				}
+			}
+		
+		return ans;
+	}
+	
+	public Qubit getHighestQubitLessThanMe(final Visualizer vis) {
+		int highest = -1;
+		Qubit ans = null;
+		
+		for (final Item i : vis.items)
+			if (i instanceof Qubit) {
+				final Qubit q = (Qubit) i;
+				if (q.id > highest && q.id < id) {
+					highest = q.id;
+					ans = q;
+				}
+			}
+		
+		return ans;
+	}
+
+	@Override
+	void onMenuEntryClick(int id, Visualizer vis) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.